### PR TITLE
feat(config): support default query params via originParams (closes #184)

### LINKED ISSUE #184: Support default query parameters in config origin
## Summary

Currently, the config `origin` property is a plain string representing the protocol and domain (e.g., `https://example.com`). It is prepended to relative URLs at runtime in `goTo` and `checkLink` steps. There is no way to specify default query parameters that should be appended to every URL constructed from the origin.

It would be useful to support a separate **`originParams`** config property — a key-value map of query parameters that get appended to any URL resolved against the origin.

## Use Case

A key motivating example is **Clerk testing token integration**. Clerk's testing mode requires a `__clerk_testing_token` query parameter on every request. Today, you'd need to manually include this on every `goTo` step URL, which is repetitive and error-prone. With `originParams`, the config handles this automatically:

```json
{
  "origin": "https://my-app.com",
  "originParams": {
    "__clerk_testing_token": "$CLERK_TESTING_TOKEN"
  }
}
```

Then a step like:

```json
{
  "goTo": "/dashboard"
}
```

Would resolve to: `https://my-app.com/dashboard?__clerk_testing_token=<token_value>`

## Current Behavior

- `origin` is a `string` in both the config schema (`config_v3.schema.json:55-58`) and the step-level schema (`goTo_v3.schema.json:33-37`).
- At runtime (`src/core/tests/goTo.ts:17-31`, `src/core/tests/checkLink.ts:16-31`), origin is simply prepended to the relative URL via string concatenation — no query parameter handling exists.

## Proposed Behavior

- Add a new `originParams` property to the config schema as an object of string key-value pairs.
- When resolving a relative URL against the origin, append the `originParams` as query parameters to the constructed URL.
- Environment variable substitution (e.g., `$CLERK_TESTING_TOKEN`) in param values would be valuable for secrets.

## Benefits

- **Clerk integration**: Set `__clerk_testing_token` once in config instead of on every step.
- **Analytics/tracking**: Append UTM or internal tracking params globally.
- **Auth tokens**: Pass API keys or session tokens via query params where needed.
- **DRY configuration**: Reduces repetition and potential for errors across test specs.

### PR DESCRIPTION
**Pull request by hawkeyexl**
## Summary

Closes #184. Adds a `config.originParams` key-value map plus step-level `params` on `goTo` and `checkLink` that are auto-appended as query parameters to any URL resolved against `origin`.

Motivating case: Clerk testing-mode requires `__clerk_testing_token` on every request. Today that has to be repeated on every step. With this change, you set it once in config and every relative-URL step picks it up.

```jsonc
{
  \"origin\": \"https://my-app.com\",
  \"originParams\": {
    \"__clerk_testing_token\": \"\$CLERK_TESTING_TOKEN\"
  }
}
```

Step like \`{\"goTo\": \"/dashboard\"}\` now resolves to \`https://my-app.com/dashboard?__clerk_testing_token=<value>\`.

## Semantics

- **Merge, not replace.** \`step.params\` is merged on top of \`config.originParams\`; step keys win on collision. Adding one per-step param does *not* silently drop a config-level token.
- **Dedupes on existing query keys.** If the resolved URL already contains a colliding key, the param here replaces it (no \`?token=old&token=new\`).
- **Fragment-safe.** A URL fragment (\`#section\`) is preserved and new params are inserted before it.
- **Env var substitution** via the existing \`\$VAR\` syntax works out of the box — \`replaceEnvs\` already runs over both config and step objects upstream.
- **Secrets warning.** Schema descriptions explicitly call out that values end up in request URLs and appear in test results, logs, and reports. Prefer short-lived testing tokens to long-lived secrets.

## Implementation

- New \`appendQueryParams(url, params)\` helper in \`src/core/utils.ts\` — fragment-aware, uses \`URLSearchParams.set\` for dedupe.
- \`goTo\` / \`checkLink\` runtime merge \`config.originParams\` with step-level \`params\` after prepending origin, then call the helper.
- Added \`originParams\` to \`config_v3\` and \`params\` to \`goTo_v3\` / \`checkLink_v3\` source schemas (+ Clerk-style examples). Output schemas and TS types regenerated via \`npm run build\`.

## Adversarial review findings addressed

An adversarial review via copilot-cli (gpt-5.2) surfaced these issues during development, all fixed:

- **[Critical] Fragment handling:** initial append was raw concat — put params inside the fragment.
- **[High] Replace-vs-merge footgun:** initial semantics were step replaces config; flipped to merge.
- **[High] Secret-in-logs visibility:** documented explicitly in all three schemas.
- **[Medium] Duplicate-key ambiguity:** now deduped via URLSearchParams.set.
- **[Medium] Dead detectTests path:** removed. With runtime merge, the earlier \`config.originParams → step.params\` copy on markup-detected steps had no behavioral effect.

Naming stays split (\`originParams\` at config, \`params\` at step) — the contexts are different enough that the inconsistency is justifiable.

## Test plan

- [x] \`test/utils.test.js\` — appendQueryParams unit tests (fragment, dedupe, null filtering, coercion)
- [x] \`test/goTo.test.js\` (new) — 6 integration cases via a stub driver
- [x] \`test/checkLink.test.js\` — 6 cases against a local HTTP server (merge, dedupe, env-var, absolute-URL skip, empty params)
- [x] \`src/common\` schema + validation suite: 575 passing
- [x] Root suite: 341 passing (+19 over baseline). Existing 16 Chrome/WebDriver/screenshot failures are environmental (no local Chrome) and present on \`main\` — unchanged from baseline.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Support for query parameters on navigation and link-checking steps via an optional params field
  * Global default query parameters via originParams configuration
  * Step-level params override global defaults on key collisions
  * Environment-variable substitution in parameter values using $VAR syntax

* **Tests**
  * Added comprehensive tests validating param merging, substitution, fragment preservation, and replacement vs. duplication behavior

### CODE REVIEW

File: test/checkLink.test.js [hunk 1]
```diff
@@ -398,6 +399,117 @@ describe("checkLink retry, headers, and HEAD fallback", function () {
...
+        },
+      },
+    });
+    assert.equal(result.status, "PASS", result.description);
+    const qs = new URLSearchParams(
+      requestLog["echo-headers-url"].split("?")[1]
+    );
+    assert.equal(qs.get("token"), "env-token-xyz");
+    delete process.env.TEST_DD_TOKEN;
+  });
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply test/checkLink.test.js:504]: Good catch — fixed in 3cced1b. The test now passes the literal `$TEST_DD_TOKEN` in `params` and calls `replaceEnvs(step)` to mirror the real runTests flow before invoking `checkLink`. Also wrapped the body in try/finally so env cleanup runs even on assertion failure.

**Comment by hawkeyexl** [comment]
Thanks @coderabbitai for the thorough pass. Responding to the outside-diff and nitpick items from your review (the inline thread is resolved separately):

**Outside diff — `step.params` dropped on absolute URLs (goTo.ts / checkLink.ts)**: Fixed via option 1. Both executors now apply `step.<action>.params` unconditionally and keep `config.originParams` gated on `isRelativeUrl`. Added a new test case in `test/goTo.test.js` that sets a per-step `params` on an absolute URL and asserts the config params do NOT leak onto it while the step params do.

**Nit — `appendQueryParams` array edge case**: Added an `Array.isArray(params)` guard. New unit test case in `test/utils.test.js` covers it.

**Nit — stubDriver false-positive risk in `test/goTo.test.js`**: Tightened. Every test now calls `assert.notEqual(calls.url, undefined, "driver.url was never invoked")` before inspecting the URL.

**Nit — missing end-to-end `$VAR` test for `goTo`**: Added. `substitutes \$VAR in originParams via replaceEnvs` in `test/goTo.test.js` passes a literal `\$TEST_DD_GOTO_TOKEN` through `replaceEnvs(config)` and asserts the substituted value lands in the URL seen by the stub driver.

**Nit — `originParams` example in `config_v3`**: Added to the examples array (Clerk-style token via `\$CLERK_TESTING_TOKEN`).

**Nit — extract merge+append helper**: Declining. The block is 4 lines duplicated across exactly 2 call sites with identical shape, and the project guidelines explicitly prefer small duplication over premature abstraction (\"Three similar lines is better than a premature abstraction\"). Happy to revisit if a third caller shows up.

**CodeQL alert on src/core/utils.ts:254 (and detectTests.ts:188)**: These fire on `spawnCommand` / `spawn` and the DITA `spawnCommand(\"dita\", [...])` call respectively. Both lines are pre-existing code unchanged by this PR — only the line numbers moved because I added `appendQueryParams` earlier in `utils.ts`. Confirmed by diffing `git show 3405c34:src/core/utils.ts` vs current. Not this PR's scope to fix.

**Comment by hawkeyexl** [comment]
@coderabbitai review

File: src/core/utils.ts [hunk 2]
```diff
@@ -36,6 +37,33 @@ function isRelativeUrl(url: string) {
...
+  // than doubled up.
+  const queryIdx = base.indexOf("?");
+  const pathAndAuthority = queryIdx >= 0 ? base.slice(0, queryIdx) : base;
+  const existingQuery = queryIdx >= 0 ? base.slice(queryIdx + 1) : "";
+  const searchParams = new URLSearchParams(existingQuery);
+  for (const [k, v] of entries) searchParams.set(k, String(v));
+
+  const query = searchParams.toString();
+  return pathAndAuthority + (query ? "?" + query : "") + fragment;
+}
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:88]: Fixed in d95991d. `appendQueryParams` now short-circuits when no new key collides with one already in the URL: it appends `&k=v` raw and preserves the existing query verbatim. Only when there's an actual collision does it fall back to `URLSearchParams` parse-set-serialize. New regression test in `test/utils.test.js` (`preserves the existing query-string verbatim when there is no key collision`) uses a typical signed-URL shape (`sig=abc:123,xyz&q=hello+world`) that would have been re-encoded before.

File: src/core/tests/goTo.ts [hunk 3]
```diff
@@ -29,6 +31,15 @@ async function goTo({ config, step, driver }: { config: any; step: any; driver:
     step.goTo.url = step.goTo.origin + step.goTo.url;
   }
 
+  // config.originParams only apply to URLs resolved against an origin;
+  // step.goTo.params applies regardless so per-step params on absolute URLs
+  // aren't silently dropped. Step keys win on collision.
+  const params = {
+    ...(relative ? config.originParams : undefined),
+    ...step.goTo.params,
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/tests/goTo.ts]: Good catch — fixed in aa4c67b. Switched to a two-pass apply: each source (`config.originParams`, then `step.goTo.params`) goes through `appendQueryParams` independently so the `Array.isArray` guard sees each source intact. Step still wins on collision because the second pass dedupes against the URL the first pass produced. New regression test in `test/goTo.test.js` (`ignores array-shaped config.originParams`) uses an array on `config.originParams` (config isn't re-validated at the runtime layer, so that's the realistic sneak-past path) and asserts the URL is unchanged.

File: src/core/tests/checkLink.ts [hunk 4]
```diff
@@ -150,6 +152,15 @@ async function checkLink({ config, step }: { config: any; step: any }) {
...
   }
 
+  // config.originParams only apply to URLs resolved against an origin;
+  // step.checkLink.params applies regardless so per-step params on absolute
+  // URLs aren't silently dropped. Step keys win on collision.
+  const params = {
+    ...(relative ? config.originParams : undefined),
+    ...step.checkLink.params,
+  };
+  step.checkLink.url = appendQueryParams(step.checkLink.url, params);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/tests/checkLink.ts]: Same fix as the goTo thread — applied in aa4c67b. `checkLink` now routes `config.originParams` and `step.checkLink.params` through `appendQueryParams` in two passes, so an array-shaped input on either source hits the `Array.isArray` guard intact instead of being smuggled through as `{0: "x"}` via spread.

File: src/core/utils.ts [hunk 5]
```diff
@@ -36,6 +37,51 @@ function isRelativeUrl(url: string) {
...
+  // URLs and strict backends. Only fall back to parse-set-serialize
+  // when we actually need dedupe.
+  const existingKeys = new Set(new URLSearchParams(existingQuery).keys());
+  const hasCollision = entries.some(([k]) => existingKeys.has(k));
+
+  let query: string;
+  if (hasCollision) {
+    const searchParams = new URLSearchParams(existingQuery);
+    for (const [k, v] of entries) searchParams.set(k, String(v));
+    query = searchParams.toString();
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Good catch — fixed in ace2d83. Unified both paths: walk the existing query and drop only the segments whose key collides with a new entry, then append the new pairs (encoded fresh). Non-colliding pairs are preserved byte-for-byte regardless of whether a collision exists. New regression test in `test/utils.test.js` (`on collision, replaces only the colliding key and keeps non-colliding pairs verbatim`) uses the same signed-URL shape as the no-collision test plus an added `token=old` collision; asserts `sig=abc:123,xyz` and `q=hello+world` come through untouched while only `token` is replaced.
